document_ai_warehouse/common/src/common/utils/document_ai_utils.py

```diff
@@ -0,0 +1,312 @@
+            return [], False
+
+        elapsed = "{:.0f}".format(time.time() - start)
+        Logger.info(f"batch_extraction - Elapsed time for operation {elapsed} seconds")
+
+        # Once the operation is complete,
+        # get output document information from operation metadata
+        metadata = documentai.BatchProcessMetadata(operation.metadata)
+        if metadata.state != documentai.BatchProcessMetadata.State.SUCCEEDED:
+            raise ValueError(
+                f"batch_extraction - Batch Process Failed: {metadata.state_message}"
+            )
+
+        documents = (
+            {}
+        )  # Contains per processed document, keys are path to original document
+
+        # One process per Input Document
+        for process in metadata.individual_process_statuses:
+            # output_gcs_destination format: gs://BUCKET/PREFIX/OPERATION_NUMBER/INPUT_FILE_NUMBER/
+            # The Cloud Storage API requires the bucket name and URI prefix separately
+            input_gcs_source = process.input_gcs_source
+            matches = re.match(r"gs://(.*?)/(.*)", process.output_gcs_destination)
+            if not matches:
+                Logger.error(
+                    f"batch_extraction - Could not parse output GCS destination:[{process.output_gcs_destination}] for {input_gcs_source}"
+                )
+                Logger.error(f"batch_extraction - {process.status}")
+                continue
+
+            output_bucket, output_prefix = matches.groups()
+            output_gcs_destination = process.output_gcs_destination
+
+            Logger.info(
+                f"batch_extraction - Handling DocAI results for {input_gcs_source} using "
+                f"process output {output_gcs_destination}"
+            )
+            # Get List of Document Objects from the Output Bucket
+            output_blobs = storage_client.list_blobs(
+                output_bucket, prefix=output_prefix + "/"
+            )
+
+            # Document AI may output multiple JSON files per source file
+            # Sharding happens when the output JSON File gets over a size threshold,
+            # like 10MB or something. I have seen it happen around 30 pages, but more
+            # often around 40 or 50 pages.
+            for blob in output_blobs:
+                # Document AI should only output JSON files to GCS
+                if ".json" not in blob.name:
+                    Logger.warning(
+                        f"batch_extraction - Skipping non-supported file: {blob.name} - Mimetype: {blob.content_type}"
+                    )
+                    continue
+
+                if input_gcs_source not in documents.keys():
+                    documents[input_gcs_source] = []
+                documents[input_gcs_source].append(f"gs://{output_bucket}/{blob.name}")
+
+        result = {}
+        for doc in documents:
+            result[doc] = merge_json_files(documents[doc])
+
+        return result
```

Comment:
Once you have the batch processing started, you can use the toolbox for the rest of this function.

https://cloud.google.com/python/docs/reference/documentai-toolbox/latest/google.cloud.documentai_toolbox.wrappers.document.Document#google_cloud_documentai_toolbox_wrappers_document_Document_from_batch_process_operation